armory/scenarios/poisoning_cifar10_witches_brew.py

```diff
@@ -0,0 +1,260 @@
+        adhoc_config = self.config.get("adhoc") or {}
+        attack_config = self.config["attack"]
+        if attack_config.get("type") == "preloaded":
+            raise ValueError("preloaded attacks not currently supported for poisoning")
+
+        self.use_poison = bool(adhoc_config["poison_dataset"])
+
+        dataset_config = self.config["dataset"]
+        test_dataset = config_loading.load_dataset(
+            dataset_config, split="test", num_batches=None, **self.dataset_kwargs
+        )
+        x_test, y_test = (np.concatenate(z, axis=0) for z in zip(*list(test_dataset)))
+
+        # TODO Needs discussion -- how are we actually going to pick trigger images.
+        trigger_index = adhoc_config["trigger_index"]
+        if isinstance(trigger_index, int):
+            trigger_index = [trigger_index]
+        self.trigger_index = trigger_index
+
+        target_class = adhoc_config["target_class"]
+        if isinstance(target_class, int):
+            target_class = [target_class] * len(self.trigger_index)
+        if len(target_class) == 1:
+            target_class = target_class * len(self.trigger_index)
+        self.target_class = target_class
+
+        source_class = adhoc_config["source_class"]
+        if isinstance(source_class, int):
+            source_class = [source_class] * len(self.trigger_index)
+        if len(source_class) == 1:
+            source_class = source_class * len(self.trigger_index)
+        self.source_class = source_class
+
+        if len(self.target_class) != len(self.trigger_index):
+            raise ValueError(
+                "target_class should have one element or be the same length as trigger_index"
+            )
+
+        if len(self.source_class) != len(self.trigger_index):
+            raise ValueError(
+                "source_class should have one element or be the same length as trigger_index"
+            )
+
+        for i, trigger_ind in enumerate(self.trigger_index):
+            if y_test[trigger_ind] != self.source_class[i]:
+                raise ValueError(
+                    f"Trigger image {i} does not belong to source class (class {y_test[trigger_ind]} != class {self.source_class[i]})"
+                )
+
+        if self.use_poison:
+
+            attack_config["kwargs"]["percent_poison"] = adhoc_config[
+                "fraction_poisoned"
+            ]
+            attack_config["kwargs"]["source_class"] = self.source_class
+            attack_config["kwargs"]["target_class"] = self.target_class
+
+            data_filepath = (
+                attack_config["kwargs"].pop("data_filepath")
+                if "data_filepath" in attack_config["kwargs"].keys()
+                else None
+            )
+
+            attack = config_loading.load_attack(attack_config, self.model)
+            if data_filepath is not None:
+                data_filepath = os.path.join(
+                    paths.runtime_paths().dataset_dir, data_filepath
+                )
```

Comment:
Let's use the models directory, actually, as I would like to keep the dataset_dir just populated by TFDS datasets.

I think we should also put it in a subdir like "attacks" as so:
```
attack_dir = os.path.join(paths.runtime_paths().saved_model_dir, "attacks")
os.makedirs(attack_dir, exist_ok=True)
data_filepath = os.path.join(attack_dir, data_filepath)
```

Reply:
Done